llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

```diff
@@ -10195,11 +10193,9 @@ BoUpSLP::getEntryCost(const TreeEntry *E, ArrayRef<Value *> VectorizedVals,
       return VecCost;
     };
     if (SLPReVec && !E->isAltShuffle())
-      return GetCostDiff(GetScalarCost, [](InstructionCost) {
-        // shufflevector will be eliminated by instcombine because the
-        // shufflevector masks are used in order (guaranteed by
-        // getShufflevectorNumGroups). The vector cost is 0.
-        return TTI::TCC_Free;
+      return GetCostDiff(GetScalarCost, [&](InstructionCost) {
+        return ::getShuffleCost(*TTI, TargetTransformInfo::SK_PermuteSingleSrc,
```

Comment:
I would recommend checking if the mask created here is an identity mask and return cost Free if it is so explicitly.